tiny_dnn/core/kernels/fully_connected_op_internal.h

```diff
@@ -12,62 +12,90 @@
 namespace tiny_dnn {
 namespace kernels {
 
-inline void fully_connected_op_internal(const tensor_t &in_data,
-                                        const vec_t &W,
-                                        const vec_t &bias,
-                                        tensor_t &out_data,
-                                        const fully_params &params,
+/**
+ * Forward propogation for fully connected layer with internal backend
+ * @param in_data
+ * @param weights
+ * @param bias
+ * @param out_data
+ * @param params
+ * @param layer_parallelize
+ */
+template <typename S1, typename S2, typename S3, typename S4>
+inline void fully_connected_op_internal(const Tensor<float_t, S1> &in_data,
+                                        const Tensor<float_t, S2> &weights,
+                                        const Tensor<float_t, S3> &bias,
+                                        Tensor<float_t, S4> &out_data,
+                                        const bool has_bias,
                                         const bool layer_parallelize) {
-  for_i(layer_parallelize, in_data.size(), [&](int sample) {
-    const vec_t &in = in_data[sample];
-    vec_t &out      = out_data[sample];
-
-    for (serial_size_t i = 0; i < params.out_size_; i++) {
-      out[i] = float_t{0};
-      for (serial_size_t c = 0; c < params.in_size_; c++) {
-        out[i] += W[c * params.out_size_ + i] * in[c];
+  size_t out_size = out_data.shape()[1], in_size = in_data.shape()[1];
+  for_i(layer_parallelize, in_data.shape()[0], [&](int sample) {
+    for (serial_size_t i = 0; i < out_size; i++) {
+      out_data.host_at(sample, i) = float_t{0};
+      for (serial_size_t c = 0; c < in_size; c++) {
+        out_data.host_at(sample, i) +=
+          weights.host_at(0, c * out_size + i) * in_data.host_at(sample, c);
       }
 
-      if (params.has_bias_) {
-        out[i] += bias[i];
+      if (has_bias) {
+        out_data.host_at(sample, i) += bias.host_at(0, i);
```

Comment:
Can we just check the size of bias ? When no bias we are receiving an empty Tensor right?